osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -0,0 +1,315 @@
+        private static bool populateValidModsForRuleset(int rulesetID, IEnumerable<APIMod> proposedMods, out List<Mod> valid)
+        {
+            valid = new List<Mod>();
+            bool proposedWereValid = true;
+
+            var ruleset = LegacyHelper.GetRulesetFromLegacyID(rulesetID);
+
+            foreach (var apiMod in proposedMods)
+            {
+                try
+                {
+                    // will throw if invalid
+                    valid.Add(apiMod.ToMod(ruleset));
+                }
+                catch
+                {
+                    proposedWereValid = false;
+                }
+            }
+
+            return proposedWereValid;
+        }
+
+        /// <summary>
+        /// Duplicates <see cref="CurrentItem"/> into the database.
+        /// </summary>
+        /// <param name="db">The database connection.</param>
+        private async Task duplicateCurrentItem(IDatabaseAccess db)
+        {
+            var newItem = new MultiplayerPlaylistItem
+            {
+                BeatmapID = CurrentItem.BeatmapID,
+                BeatmapChecksum = CurrentItem.BeatmapChecksum,
+                RulesetID = CurrentItem.RulesetID,
+                AllowedMods = CurrentItem.AllowedMods,
+                RequiredMods = CurrentItem.RequiredMods
+            };
+
+            newItem.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, newItem));
+            await hub.OnPlaylistItemAdded(room, newItem);
+        }
+
+        /// <summary>
+        /// Updates <see cref="CurrentItem"/> and the playlist item ID stored in the room's settings.
+        /// </summary>
+        /// <param name="notifyHub">Whether to notify the <see cref="MultiplayerHub"/> of the change.</param>
+        private async Task updateCurrentItem(bool notifyHub = true)
+        {
+            using (var db = dbFactory.GetInstance())
+            {
+                var allItems = await db.GetAllPlaylistItems(room.RoomID);
+
+                switch (mode)
+                {
+                    default:
+                        // Pick the first available non-expired playlist item, or default to the last item for when all items are expired.
+                        CurrentItem = await (allItems.FirstOrDefault(i => !i.expired) ?? allItems.Last()).ToMultiplayerPlaylistItem(db);
+                        break;
+
+                    case QueueModes.FairRotate:
+                        // Todo: Group playlist items by (user_id -> count_expired), and select the first available playlist item from a user that has available beatmaps where count_expired is the lowest.
+                        throw new NotImplementedException();
+                }
+            }
+
+            long lastItem = room.Settings.PlaylistItemId;
+            room.Settings.PlaylistItemId = CurrentItem.ID;
+
+            if (notifyHub && CurrentItem.ID != lastItem)
```

Comment:
Is there a reason to block this? If it's just an optimisation then I don't think it's required (in the interest of keeping things as simple as possible).

Reply:
If I remember correctly the only call that doesn't notify is in `Initialise`. In which case it makes sense to not notify, but maybe this method can be factored differently as to not require the flag. Not sure.

Reply:
Why does it make sense? Does the notify trigger something client-side which we want to avoid? My thinking was that it would be a no-op therefore there's no real reason to block it.

Reply:
Not a huge deal either way, just looking for any way to reduce complexity. Can be left in place if it's deemed best for the server-client flow.

Reply:
Removed. It's just a no-op instead.